File: depackio/LZSS_decompress.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

#define N_BITS 12
#define F_BITS 4
#define THRESHOLD 2
#define N (1 << N_BITS)
#define F ((1 << F_BITS) + THRESHOLD)

// Decoder state
struct LZSSDState_t
{
  const uint8_t *input_ptr;
  uint16_t input_pos;
  uint16_t input_size;
  uint8_t *output_ptr;
  uint16_t output_size;
  uint8_t *dict_ptr;
  uint8_t last_ch;
};

static bool LZSS_GetChar (struct LZSSDState_t *ds)
{
  if (ds->input_pos < ds->input_size)
  {
    ds->last_ch = ds->input_ptr[ds->input_pos++];
    return false;
  }
  else
    return true;
}

static void LZSS_PutChar (struct LZSSDState_t *ds)
{
  ds->output_ptr[ds->output_size++] = ds->last_ch;
}

static void LZSS_decode (struct LZSSDState_t *ds)
{
  uint16_t d = 0;
  uint32_t j = N - F;

  ds->input_pos = 0;
  ds->output_size = 0;

  for (;;)
  {
    d >>= 1;
    if ((d & 0xFF00) == 0)
    {
      if (LZSS_GetChar (ds)) break;
      d = 0xFF00 + ds->last_ch;
    }
    if ((d & 1) != 0)
    {
      if (LZSS_GetChar (ds)) break;
      ds->dict_ptr[j++] = ds->last_ch;
      j &= N - 1;
      LZSS_PutChar (ds);
    }
    else
    {
      uint32_t i;
      uint8_t a, b;

      if (LZSS_GetChar (ds)) break;
      b = ds->last_ch;
      if (LZSS_GetChar (ds)) break;
      i = ((ds->last_ch << 4) & 0xFF00) + b;
      a = (ds->last_ch & 0x0F) + THRESHOLD + 1;
      do
      {
        i &= N - 1;
        ds->last_ch = ds->dict_ptr[i++];
        ds->dict_ptr[j++] = ds->last_ch;
        j &= N - 1;
        LZSS_PutChar (ds);
      } while (--a);
    }
  }
}

static struct LZSSDState_t lzss_ds;

uint16_t LZSS_decompress (const void *source, void *dest, uint16_t size)
{
  memset (work_mem, 0, WORKMEM_SIZE);
  lzss_ds.input_ptr = source;
  lzss_ds.input_size = size;
  lzss_ds.output_ptr = dest;
  lzss_ds.dict_ptr = work_mem;
  LZSS_decode (&lzss_ds);
  return lzss_ds.output_size;
}

File: adt2pack/pas/LZSS_decompress.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

const
  N_BITS = 12;
  F_BITS = 4;
  THRESHOLD = 2;
  N = 1 SHL N_BITS;
  F = (1 SHL F_BITS)+THRESHOLD;

procedure LZSS_decode;

label
  j1,j2,j3,j4,j5;

var
  al,cl,ch,cf: Byte;
  dx: Word;
  ebx,edi: Dword;

procedure GetChar;
begin
  If (ibufCount < ibufSize) then
    begin
      al := input_ptr[ibufCount];
      Inc(ibufCount);
      cf := 0;
    end
  else
    cf := 1;
end;

procedure PutChar;
begin
  output_ptr[output_size] := al;
  Inc(output_size);
end;

begin
    ibufCount := 0;
    ibufSize := input_size;
    output_size := 0;
    ebx := 0;                        //        xor     ebx,ebx
    dx := 0;                         //        xor     edx,edx
    edi := N-F;                      //        mov     edi,N-F
j1: dx := dx SHR 1;                  //@@1:    shr     dx,1
    If (dx SHR 8 <> 0) then          //        or      dh,dh
      GOTO j2;                       //        jnz     @@2
    GetChar;                         //        call    GetChar
    If (cf = 1) then GOTO j5;        //        jc      @@5
    dx := $ff00 OR al;               //        mov     dh,0ffh
                                     //        mov     dl,al
j2: If (dx AND 1 = 0) then           //@@2:    test    dx,1
      GOTO j3;                       //        jz      @@3
    GetChar;                         //        call    GetChar
    If (cf = 1) then GOTO j5;        //        jc      @@5
                                     //        push    esi
    work_ptr[edi] := al;             //        mov     esi,work_ptr
                                     //        add     esi,edi
                                     //        mov     byte ptr [esi],al
                                     //        pop     esi
    edi := (edi+1) AND (N-1);        //        inc     edi
                                     //        and     edi,N-1
    PutChar;                         //        call    PutChar
    GOTO j1;                         //        jmp     @@1
j3: GetChar;                         //@@3:    call    GetChar
    If (cf = 1) then GOTO j5;        //        jc      @@5
    ch := al;                        //        mov     ch,al
    GetChar;                         //        call    GetChar
    If (cf = 1) then GOTO j5;        //        jc      @@5
                                     //        mov     bh,al
                                     //        mov     cl,4
    ebx := (al SHL 4) AND $ff00;     //        shr     bh,cl
    ebx := ebx OR ch;                //        mov     bl,ch
                                     //        mov     cl,al
                                     //        and     cl,0fh
    cl := (al AND $0f)+THRESHOLD;    //        add     cl,THRESHOLD
    Inc(cl);                         //        inc     cl
j4: ebx := ebx AND (N-1);            //@@4:    and     ebx,N-1
                                     //        push    esi
    al := work_ptr[ebx];             //        mov     esi,work_ptr
                                     //        mov     al,byte ptr [esi+ebx]
                                     //        add     esi,edi
    work_ptr[edi] := al;             //        mov     byte ptr [esi],al
                                     //        pop     esi
    Inc(edi);                        //        inc     edi
    edi := edi AND (N-1);            //        and     edi,N-1
    PutChar;                         //        call    PutChar
    Inc(ebx);                        //        inc     ebx
    Dec(cl);                         //        dec     cl
    If (cl <> 0) then GOTO j4;       //        jnz     @@4
    GOTO j1;                         //        jmp     @@1
j5:                                  //@@5:
end;

function LZSS_decompress(var source,dest; size: Word): Word;

begin
  input_ptr := @source;
  output_ptr := @dest;
  work_ptr := @work_mem;
  input_size := size;
  FillChar(work_ptr^,WORKMEM_SIZE,0);
  LZSS_decode;
  LZSS_decompress := output_size;
end;
